torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2287,6 +2250,7 @@ def _load_sharded_state_dict(
     def forward(self, *args: Any, **kwargs: Any) -> Any:
         with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):
             self._lazy_init()
+            self._wait_for_previous_optim_step()
```

Comment:
I separated out `_wait_for_previous_optim_step()`, which performs *synchronization*, from `_lazy_init()` since it does not logically match.

`_lazy_init()` has been refactored to only run its actual logic once, meaning that it is a no-op after the first `forward()`. Before, the implementation still relied on the `_wait_for_previous_optim_step()` synchronization even though the actual initialization logic in `_lazy_init()` was a no-op.

Other methods like `state_dict()` and `summon_full_params()` call `_lazy_init()` but require hard synchronization using `torch.cuda.synchronize()`, so those do not need the additional `_wait_for_previous_optim_step()` call, further motivating the separation.